components/user_groups_modal/user_groups_modal.tsx

```diff
@@ -0,0 +1,469 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {createRef, RefObject} from 'react';
+
+import {Modal} from 'react-bootstrap';
+
+import {FormattedMessage} from 'react-intl';
+
+import Constants, {ModalIdentifiers} from 'utils/constants';
+
+import FaSearchIcon from 'components/widgets/icons/fa_search_icon';
+import * as Utils from 'utils/utils.jsx';
+import {Group, GroupSearachParams, GroupPermissions} from 'mattermost-redux/types/groups';
+
+import './user_groups_modal.scss';
+import MenuWrapper from 'components/widgets/menu/menu_wrapper';
+import Menu from 'components/widgets/menu/menu';
+import {ModalData} from 'types/actions';
+import {debounce} from 'mattermost-redux/actions/helpers';
+import Input from 'components/input';
+import CreateUserGroupsModal from 'components/create_user_groups_modal';
+import LoadingScreen from 'components/loading_screen';
+import ViewUserGroupModal from 'components/view_user_group_modal';
+import {ActionResult} from 'mattermost-redux/types/actions';
+
+const GROUPS_PER_PAGE = 60;
+
+export type Props = {
+    onExited: () => void;
+    groups: Group[];
+    myGroups: Group[];
+    searchTerm: string;
+    currentUserId: string;
+    backButtonAction: () => void;
+    groupPermissionsMap: Record<string, GroupPermissions>;
+    actions: {
+        getGroups: (
+            filterAllowReference?: boolean,
+            page?: number,
+            perPage?: number,
+            includeMemberCount?: boolean
+        ) => Promise<{data: Group[]}>;
+        setModalSearchTerm: (term: string) => void;
+        getGroupsByUserIdPaginated: (
+            userId: string,
+            filterAllowReference?: boolean,
+            page?: number,
+            perPage?: number,
+            includeMemberCount?: boolean
+        ) => Promise<{data: Group[]}>;
+        openModal: <P>(modalData: ModalData<P>) => void;
+        searchGroups: (
+            params: GroupSearachParams,
+        ) => Promise<{data: Group[]}>;
+        removeUsersFromGroup: (groupId: string, userIds: string[]) => Promise<ActionResult>;
+        addUsersToGroup: (groupId: string, userIds: string[]) => Promise<ActionResult>;
+        archiveGroup: (groupId: string) => Promise<ActionResult>;
+    };
+}
+
+type State = {
+    page: number;
+    myGroupsPage: number;
+    loading: boolean;
+    show: boolean;
+    selectedFilter: string;
+    allGroupsFull: boolean;
+    myGroupsFull: boolean;
+}
+
+export default class UserGroupsModal extends React.PureComponent<Props, State> {
```

Comment:
This component is huge, can we split this in subcomponents (probably functional components)?

Reply:
Moved the modal header and group list out into their own smaller functional components. Reduced the size of this file by about 100 LOC